src/highdicom/sr/content.py

```diff
@@ -599,3 +838,163 @@ def __init__(
+             use the information in the Referenced Series Sequence, which
+             applies to all segments and frames present in the segmentation
+             instance.
+
+        """
+        if dataset.SOPClassUID != SegmentationStorage:
+            raise ValueError(
+                'Input dataset should be a segmentation storage instance'
+            )
+
+        source_images = []
+        source_series = None
+
+        # Attempt to retrieve a list of source images for the specific segment
+        # and frames specified using information in the per-frame functional
+        # groups sequence
+        # This gives more specific information, but depends upon optional
+        # attributes so may fail with some segmentation objects
+        # Those created with highdicom should have this information
+        referenced_frame_info = []
+        if frame_numbers is not None:
+            # Use only the provided frames
+            for f in frame_numbers:
+                # Check that the provided frame number is valid
+                if f < 0 or f >= dataset.NumberOfFrames:
+                    raise ValueError(
+                        f'Frame {f} is an invalid frame number within the '
+                        'provided dataset'
+                    )
+
+                frame_info = dataset.PerFrameFunctionalGroupsSequence[f]
+
+                # Check that this frame references the correct
+                # segment
+                ref_segment = frame_info.SegmentIdentificationSequence[0]\
+                    .ReferencedSegmentNumber
+
+                if ref_segment != segment_number:
+                    raise ValueError(
+                        f'The provided frame number {f} does not refer to '
+                        f'segment number {segment_number}'
+                    )
+                referenced_frame_info.append(frame_info)
+        else:
+            # Use every frame that applies to this segment
+            for frame_info in dataset.PerFrameFunctionalGroupsSequence:
+                ref_segment = frame_info.SegmentIdentificationSequence[0]\
+                    .ReferencedSegmentNumber
+                if ref_segment == segment_number:
+                    referenced_frame_info.append(frame_info)
+
+            if len(referenced_frame_info) == 0:
+                raise ValueError(
+                    f'No frame information found referencing segment '
+                    f'{segment_number}'
+                )
+
+        # Gather up references using the Derivation Image Sequences of the
+        # referenced frames
+        refd_insuids = set()
+        for frame_info in referenced_frame_info:
+            if hasattr(frame_info, 'DerivationImageSequence'):
+                for drv_image in frame_info.DerivationImageSequence:
+                    if hasattr(drv_image, 'SourceImageSequence'):
+                        for src_image in drv_image.SourceImageSequence:
```

Comment:
The two lines could be combined into one:

```python
for src_image in getattr(drv_image, 'SourceImageSequence', []):
    ...
```

Reply:
Addressed in 4bea49d